src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker_Generators.cs

```diff
@@ -0,0 +1,159 @@
+                solution,
+                this.ProjectState.Id,
+                (service, solutionChecksum, cancellationToken) => service.GetContentsAsync(
+                    solutionChecksum, projectId, documentsToAddOrUpdate.ToImmutable(), cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!generatedSourcesOpt.HasValue)
+                return null;
+
+            var generatedSources = generatedSourcesOpt.Value;
+            Contract.ThrowIfTrue(generatedSources.Length != documentsToAddOrUpdate.Count);
+
+            // Now go through and produce the new document states, using what we have already if it is unchanged, or
+            // what we have retrieved for anything new/changed.
+            using var generatedDocumentsBuilder = TemporaryArray<SourceGeneratedDocumentState>.Empty;
+            foreach (var (documentIdentity, contentIdentity) in infos)
+            {
+                var addOrUpdateIndex = documentsToAddOrUpdate.IndexOf(documentIdentity.DocumentId);
+                if (addOrUpdateIndex >= 0)
+                {
+                    // a document whose content we fetched from the remote side.
+                    var generatedSource = generatedSources[addOrUpdateIndex];
+                    var sourceText = SourceText.From(
+                        generatedSource, contentIdentity.EncodingName is null ? null : Encoding.GetEncoding(contentIdentity.EncodingName), contentIdentity.ChecksumAlgorithm);
+
+                    var generatedDocument = SourceGeneratedDocumentState.Create(
+                        documentIdentity,
+                        sourceText,
+                        ProjectState.ParseOptions!,
+                        ProjectState.LanguageServices);
+                    Contract.ThrowIfTrue(generatedDocument.GetTextChecksum() != contentIdentity.Checksum, "Checksums must match!");
+                    generatedDocumentsBuilder.Add(generatedDocument);
+                }
+                else
+                {
+                    // a document that already matched something locally.
+                    var existingDocument = generatorInfo.Documents.GetRequiredState(documentIdentity.DocumentId);
+                    Contract.ThrowIfTrue(existingDocument.Identity != documentIdentity, "Identies must match!");
+                    Contract.ThrowIfTrue(existingDocument.GetTextChecksum() != contentIdentity.Checksum, "Checksums must match!");
+                    generatedDocumentsBuilder.Add(existingDocument);
+                }
+            }
+
+            var generatedDocuments = new TextDocumentStates<SourceGeneratedDocumentState>(generatedDocumentsBuilder.ToImmutableAndClear());
+            var compilationWithGeneratedFiles = compilationWithoutGeneratedFiles.AddSyntaxTrees(
+                await generatedDocuments.States.Values.SelectAsArrayAsync(state => state.GetSyntaxTreeAsync(cancellationToken)).ConfigureAwait(false));
```

Comment:
>SelectAsArrayAsync [](http://example.com/codeflow?start=55&length=18)

I'm curious what the SelectAsArray gets you here over just Select. AddSyntaxTrees appears to take in an IEnumerable, and the SelectAsArray code will allocate an ImmutableArray whereas Select doesn't allocate an array.

Reply:
Oh, I think I missed the async part of this.

Reply:
i'm not sure if we have a SleectAsync, and was keeping this in sync with the code that runs in-process :)

I also have an afinity for the ImmutableArray versions because i know they only run onces (whereas IEnumerables scare me).